core: Adds __add__ to RequestOutput

a-ovchinnikov opened the pull request:
RequestOutput is a proper class instances of which could be
merged with each other to produce new instances of the same
class, yet the logic for merging these objects lives in a
utility function which while being dubious from design PoV
has also caused issues with circular dependencies in the past.

This commit introduces \_\_add\_\_ to RequestOutput by essentially
placing merge_outputs finctionality into the class. Now
two outputs could be merged directly:

    output1 + output2

while a list of outputs could be merged like this:

    sum(outputs)

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

eskultety reviewed (commented):
Much cleaner, thank you. Minor comments provided, but LGTM overall.

eskultety commented on hermeto/core/models/output.py:
```diff
@@ -159,6 +160,39 @@ def generate_sbom(self) -> Sbom:
+        #    for output in outputs:
+        # Since options are a dict I am opting for simply merging it here,
+        # however this is prone to the same problem and might result in
+        # different results depending on the order of argument. This has to be
```
> `options` look a bit like something that was originally thought to be important, but ended up not really widely used

Not yet I'd say - that's why it hasn't been enabled for all backends globally IIRC. Its existence is absolutely paramount though in enabling some features the adoption of which we'll see more of in the future but for which we didn't have any other alternative back then because the only other way of doing this would have been passing it via a static config YAML file which is incredibly tedious in some setups/deployments (and also because our config file should rather not exist at all in its current shape without a major rework).
Passthrough <tool> options is also something that people may likely start asking for with more adoption.

NB that not even configuration environment variables (if we had any) could help here because the set of passthrough config options is practically infinite.

eskultety commented on hermeto/core/models/output.py:
```diff
@@ -159,6 +160,39 @@ def generate_sbom(self) -> Sbom:
+        options: Optional[dict]
+        if lhs is not None and rhs is not None:
+            options = deepcopy(lhs)
+            options.update(rhs)
```
> it either must eventually be represented by a less generic structure than a dictionary

This is going to be quite complicated. While having a formal schema would be ideal, no dispute, the reality of passthrough options is that the set of options we'll have to ultimately support this way is so big that you'll end up with a generic schema anyway (my opinion).

> or we should just live with this

Fine by me for the time being until we have a better picture and more data/requests for more options.

I'm sorry, you updated the `else` clause below with deepcopy, but this merge was left intact. What's the difference between those 2 in terms of not taking the cautious approach with all operands? What am I missing?

^This is pretty much the only question I have and then we're good here.

a-ovchinnikov replied:
Ok, fair enough, will add a deepcopy for rhs too.

a-ovchinnikov commented:
In the latest revision I updated the logic for picking correct options when one of them is None, also switched to working with deepcopies of both options in order to prevent possible cross-talk.

a-ovchinnikov commented:
New revision: minor cosmetic changes.
